pkg/lib/accountmanagement/service.go

```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
 
 	return &FinishAddingOutput{}, nil
 }
+
+type ResendOTPCodeInput struct {
+	UserID       string
+	Channel      model.AuthenticatorOOBChannel
+	LoginID      string
+	isSwitchPage bool
+}
+
+func (s *Service) ResendOTPCode(input *ResendOTPCodeInput) (err error) {
+	// Either it is a switch page or resend
+	isResend := !input.isSwitchPage
+	err = s.sendOTPCode(
+		input.UserID,
+		input.Channel,
+		input.LoginID,
+		isResend,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type ChangePasswordInput struct {
+	OAuthSessionID string
+	RedirectURI    string
+	OldPassword    string
+	NewPassword    string
+}
+
+type ChangePasswordOutput struct {
+	RedirectURI string
+}
+
+// If have OAuthSessionID, it means the user is changing password after login with SDK.
+// Then do special handling such as authenticationInfo
+func (s *Service) ChangePassword(resolvedSession session.ResolvedSession, input *ChangePasswordInput) (*ChangePasswordOutput, error) {
+	userID := resolvedSession.GetAuthenticationInfo().UserID
+	redirectURI := input.RedirectURI
+
+	err := s.Database.WithTx(func() error {
+		ais, err := s.Authenticators.List(
+			userID,
+			authenticator.KeepType(model.AuthenticatorTypePassword),
+			authenticator.KeepKind(authenticator.KindPrimary),
+		)
+		if err != nil {
+			return err
+		}
+		if len(ais) == 0 {
+			return api.ErrNoPassword
+		}
+		oldInfo := ais[0]
+		_, err = s.Authenticators.VerifyWithSpec(oldInfo, &authenticator.Spec{
+			Password: &authenticator.PasswordSpec{
+				PlainPassword: input.OldPassword,
+			},
+		}, nil)
+		if err != nil {
+			err = api.ErrInvalidCredentials
+			return err
+		}
+		changed, newInfo, err := s.Authenticators.UpdatePassword(oldInfo, &service.UpdatePasswordOptions{
+			SetPassword:    true,
+			PlainPassword:  input.NewPassword,
+			SetExpireAfter: true,
+		})
+		if err != nil {
+			return err
+		}
+		if changed {
+			err = s.Authenticators.Update(newInfo)
+			if err != nil {
+				return err
+			}
+		}
+		// If is changing password with SDK.
+		if input.OAuthSessionID != "" {
+			authInfo := resolvedSession.GetAuthenticationInfo()
+			authenticationInfoEntry := authenticationinfo.NewEntry(authInfo, input.OAuthSessionID, "")
+
+			err = s.AuthenticationInfoService.Save(authenticationInfoEntry)
+			if err != nil {
+				return err
+			}
+			redirectURI = s.UIInfoResolver.SetAuthenticationInfoInQuery(input.RedirectURI, authenticationInfoEntry)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &ChangePasswordOutput{RedirectURI: redirectURI}, nil
+
+}
+
+type CreateAdditionalPasswordInput struct {
+	NewAuthenticatorID string
+	UserID             string
+	Password           string
+}
+
+func NewCreateAdditionalPasswordInput(userID string, password string) CreateAdditionalPasswordInput {
+	return CreateAdditionalPasswordInput{
+		NewAuthenticatorID: uuid.New(),
+		UserID:             userID,
+		Password:           password,
+	}
+}
```

Comment:
Remove this function. CreateAdditionalPasswordInput is simple struct that does not need a function to initialize.